src/macros.rs

```diff
@@ -66,36 +93,66 @@ macro_rules! unsafe_impl {
     ) => {
         unsafe_impl!(
             @inner
             @const $constname: $constty,
             $($tyvar $(: $(? $optbound +)* + $($bound +)*)?,)*
-            => $trait for $ty
+            => $trait for $ty $(; |$candidate $(: &$ref_repr)? $(: NonNull<$ptr_repr>)?| $is_bit_valid)?
         );
     };
     (
         $($tyvar:ident $(: $(? $optbound:ident $(+)?)* $($bound:ident $(+)?)* )?),*
-        => $trait:ident for $ty:ty
+        => $trait:ident for $ty:ty $(; |$candidate:ident $(: &$ref_repr:ty)? $(: NonNull<$ptr_repr:ty>)?| $is_bit_valid:expr)?
     ) => {
         unsafe_impl!(
             @inner
             $($tyvar $(: $(? $optbound +)* + $($bound +)*)?,)*
-            => $trait for $ty
+            => $trait for $ty $(; |$candidate $(: &$ref_repr)? $(: NonNull<$ptr_repr>)?| $is_bit_valid)?
         );
     };
     (
         @inner
         $(@const $constname:ident : $constty:ident,)*
         $($tyvar:ident $(: $(? $optbound:ident +)* + $($bound:ident +)* )?,)*
-        => $trait:ident for $ty:ty
+        => $trait:ident for $ty:ty $(; |$candidate:ident $(: &$ref_repr:ty)? $(: NonNull<$ptr_repr:ty>)?| $is_bit_valid:expr)?
     ) => {
         unsafe impl<$(const $constname: $constty,)* $($tyvar $(: $(? $optbound +)* $($bound +)*)?),*> $trait for $ty {
-            #[allow(clippy::missing_inline_in_public_items)]
-            fn only_derive_is_allowed_to_implement_this_trait() {}
+            unsafe_impl!(@method $trait $(; |$candidate: $(&$ref_repr)? $(NonNull<$ptr_repr>)?| $is_bit_valid)?);
+        }
+    };
+
+    (@method TryFromBytes ; |$candidate:ident: &$repr:ty| $is_bit_valid:expr) => {
+        #[inline]
+        unsafe fn is_bit_valid(candidate: NonNull<Self>) -> bool {
+            // SAFETY: The caller has promised that it is sound to perform this
+            // pointer cast and dereference.
+            #[allow(clippy::as_conversions)]
+            let $candidate = unsafe { &*(candidate.as_ptr() as *const $repr) };
+            $is_bit_valid
         }
     };
+    (@method TryFromBytes ; |$candidate:ident: NonNull<$repr:ty>| $is_bit_valid:expr) => {
+        #[inline]
+        unsafe fn is_bit_valid(candidate: NonNull<Self>) -> bool {
+            // SAFETY: `candidate` is a non-null pointer. The caller has
+            // promised that it is sound to invoke their `$is_bit_valid` with
+            // this pointer.
+            #[allow(clippy::as_conversions)]
+            let $candidate = unsafe { NonNull::new_unchecked(candidate.as_ptr() as *mut $repr) };
+            $is_bit_valid
+        }
+    };
+    (@method TryFromBytes) => { #[inline(always)] unsafe fn is_bit_valid(_: NonNull<Self>) -> bool { true } };
+    (@method $trait:ident) => {
+        #[allow(clippy::missing_inline_in_public_items)]
+        fn only_derive_is_allowed_to_implement_this_trait() {}
+    };
+    // TODO: Test this with trybuild.
```

Comment:
Nevermind, that's overkill. Even if this were buggy, you'd just end up with a worse compiler error message - either a  missing `is_bit_valid` impl (for `TryFromBytes`) or an unexpected `is_bit_valid` impl (for any other trait)